sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/datastore/V1ReadIT.java

```diff
@@ -92,9 +96,37 @@ public void testE2EV1Read() throws Exception {
     p.run();
   }
 
+  /**
+   * An end-to-end test for {@link DatastoreV1.Read#withGqlQuery(String)}.
+   *
+   * <p>Write some test entities to datastore and then run a dataflow pipeline that
+   * reads and counts the total number of entities. Verify that the count matches
+   * the number of entities written.
+   */
+  @Test
+  public void testE2EV1ReadWithGQLQuery() throws Exception {
+    // Read from datastore
+    String gqlQuery = String.format("SELECT * from %s WHERE __key__ HAS ANCESTOR KEY(%s, '%s')",
+        options.getKind(), options.getKind(), ancestor);
+
+    DatastoreV1.Read read = DatastoreIO.v1().read()
+        .withProjectId(project)
+        .withGqlQuery(gqlQuery)
+        .withNamespace(options.getNamespace());
+
+    // Count the total number of entities
+    Pipeline p = Pipeline.create(options);
+    PCollection<Long> count = p
+        .apply(read)
+        .apply(Count.<Entity>globally());
+
+    PAssert.thatSingleton(count).isEqualTo(numEntities);
+    p.run();
```

Comment:
We should also create tests for: Creating a template with these sources, then creating a job from that template, and making sure that works too.